bigbite/themer
# [BBT-88] Spacing UI component

## Description



Completes [BBT-88](https://b5ecom.atlassian.net/browse/BBT-88) - This PR adds a reusable 'Spacing' component that allows you to control the block gap, margin & padding styles for individual blocks.

@spbuckle started this ticket using the `__experimentalSpacingSizesControl` component, but I found it really difficult to work with and it lacks documentation. I switched to using the [`BoxControl`](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-boxcontrol--docs) component instead and I think this is still sufficient for our needs.

I've added some helpers to convert css vars to custom spacing values & vice versa. Weirdly, the WP function that converts a var to value requires you to use a vertical bar syntax (e.g. `var:preset|spacing|20`) so I've built a conversion into the helper. For converting values to vars, I tried to use [`getSpacingCssVar`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-block-editor/#getspacingpresetcssvar) but I couldn't get that to work whatsoever and ended up using my own implementation.

I've also had to bump the version of the `@wordpress/block-editor` package as the previous version didn't include some functions I needed.

## Change Log



- Added Spacing UI component.
- Added CSS length unit helper (also in other PR's)
- Added `varToSpacing` helper function
- Added `spacingToVar` helper function
- Bumped `@wordpress/block-editor` package to `12.11.0`

## Steps to test



- Add some custom spacing sizes to your `theme.json`:
```json
"settings": {
    "spacing": {
      "spacingSizes": [
				{
					"size": "3rem",
					"slug": "block-gap",
					"name": "1"
				},
        {
					"size": "5rem",
					"slug": "block-gap-lg",
					"name": "2"
				}
			]
    }
  },
```
- Add some spacing styles to a block in your `theme.json`:
```json
"styles": {
		"blocks": {
			"core/paragraph": {
				"spacing": {
					"blockGap": "var(--wp--preset--spacing--block-gap-lg)"
				}
			}
		}
	}
```
- Check that your value is imported correctly when you load themer
- Play about with different values in the margin, padding & block gap.
- Change the values to use one of your custom spacing sizes.
- Export the theme.json and check the values have been converted to vars.

## Screenshots/Videos



### Values imported correctly.

https://github.com/bigbite/themer/assets/45216035/3fcd51a3-ea32-431c-bfa5-c13df3f77c24

### Changing values 

https://github.com/bigbite/themer/assets/45216035/ff9ab85d-3f3c-440b-9a2d-2d71921d80ea

### Exporting back into theme.json

https://github.com/bigbite/themer/assets/45216035/9220dc8e-c055-458f-ad0a-ed8958ea6ffb

## Checklist:





[BBT-88]: https://b5ecom.atlassian.net/browse/BBT-88?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
